packages/cw721/src/traits.rs

```diff
@@ -1,168 +1,64 @@
-use serde::de::DeserializeOwned;
-use serde::Serialize;
-
-use crate::{
-    AllNftInfoResponse, ApprovalResponse, ApprovalsResponse, ContractInfoResponse, NftInfoResponse,
-    NumTokensResponse, OperatorResponse, OperatorsResponse, OwnerOfResponse, TokensResponse,
-};
-use cosmwasm_std::{Binary, CustomMsg, Deps, DepsMut, Env, MessageInfo, Response, StdResult};
-use cw_utils::Expiration;
-
-pub trait Cw721<T, C>: Cw721Execute<T, C> + Cw721Query<T>
-where
-    T: Serialize + DeserializeOwned + Clone,
-    C: CustomMsg,
-{
-}
-
-pub trait Cw721Execute<T, C>
-where
-    T: Serialize + DeserializeOwned + Clone,
-    C: CustomMsg,
-{
-    type Err: ToString;
-
-    fn transfer_nft(
+use std::fmt::Debug;
+
+use cosmwasm_std::{Deps, Empty, Env, MessageInfo};
+use schemars::JsonSchema;
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::error::Cw721ContractError;
+
+/// This is an exact copy of `CustomMsg`, since implementing a trait for a type from another crate is not possible.
+///
+/// Possible:
+/// `impl<T> Cw721CustomMsg for Option<T> where T: Cw721CustomMsg {}`
+///
+/// Not possible:
+/// `impl<T> CustomMsg for Option<T> where T: CustomMsg {}`
+///
+/// This will be removed once the `CustomMsg` trait is moved to the `cosmwasm_std` crate: https://github.com/CosmWasm/cosmwasm/issues/2056
+pub trait Cw721CustomMsg: Serialize + Clone + Debug + PartialEq + JsonSchema {}
+
+pub trait Cw721State: Serialize + DeserializeOwned + Clone + Debug {}
+
+impl Cw721State for Empty {}
+impl<T> Cw721State for Option<T> where T: Cw721State {}
+impl Cw721CustomMsg for Empty {}
+impl<T> Cw721CustomMsg for Option<T> where T: Cw721CustomMsg {}
+
+pub trait StateFactory<TState> {
```

Comment:
Introducing new StateFactory:

```rs
pub trait StateFactory<TState> {
    fn create(
        &self,
        deps: Option<Deps>,
        env: Option<&Env>,
        info: Option<&MessageInfo>,
        current: Option<&TState>,
    ) -> Result<TState, Cw721ContractError>;
    fn validate(
        &self,
        deps: Option<Deps>,
        env: Option<&Env>,
        info: Option<&MessageInfo>,
        current: Option<&TState>,
    ) -> Result<(), Cw721ContractError>;
}
```

Why is this useful? Before (crud) logic are spread over in `instantiate`, `execute`, and `migrate` endpoints. This is errorprone, since logic is all over the places - like there may be inconsistent validation, ownership checks/authorization, etc.

Aggregating all crud logic for each atomic state into a StateFactory improves maintenance of such code.

Reply:
What is the reasoning for calling it `StateFactory`? Does it follow the factory pattern where it is responsible for creating other data types or instantiating contracts?

Reply:
Correct. As stated by GoF for `Factory method` pattern:

> "Define an interface for creating an object, but let subclasses decide which class to instantiate. The Factory method lets a class defer instantiation it uses to subclasses."

So it is a trait implemented by various state/storage objects like: CollectionInfo, CollectionExtension, NftInfo, NftExtension (metadata), Trait, RoyaltyInfo, ...

It takes care of object creation and update. So it can be used in instantiate, execute, and migrate endpoints.

Reply:
@shanev what about changing it from `StateFactory` -> `StateController`?

Since it also allows optional authorization checks, like whether sender is creator or minter, e.g. for CollectionInfo and NftInfo.